ENH: plot method accessors

shoyer opened the pull request:
Fixes #9124

This PR adds plotting sub-methods like `df.plot.scatter()` as an alternative to using `df.plot(kind='scatter')`.

I've added meaningful function signatures and documentation for a few of these methods, but I would greatly appreciate help to fill in the rest -- this is a lot of documentation to assemble/reconstruct! The entire point of this PR, of course, is to have better introspection and docstrings.

Todo list:
- [x] Basic docstrings/signatures
  - [x] `area`
  - [x] `line`
  - [x] `bar`
  - [x] `barh`
  - [x] `box`
  - [x] `hexbin`
  - [x] `hist`
  - [x] `kde`/`density`
  - [x] `pie`
  - [x] `scatter`
- [x] Write tests for the methods
- [x] Fix groupby plots (tests currently failing)
- [x] Plotting docs
- [x] API docs
- [x] Release notes

TomAugspurger commented:
I'll do a PR against your branch for the rest of the docstrings this Saturday unless someone else gets there first.

TomAugspurger commented:
Thanks for doing this!

sinhrks commented:
Current policy is nice for docstring thus for users, but I feel this makes consistent maintenance little difficult, because maintainer have to care all the separated functions. I think it is better to have a table which organizes what options are available on every plotting function. Maybe on Wiki?

I'm willing to do, but it takes a while as each function have options not listed in the signature, for example,
https://github.com/pydata/pandas/blob/master/pandas/tools/plotting.py#L1428

shoyer commented:
@sinhrks Agreed, this is going to be tricky. For a first draft (this PR), I think use of `**kwargs` for some options is unavoidable (every function will have `**kwargs` in the signature). Eventually, we can try to stop using it but practically that may not be possible.

TomAugspurger commented:
I hope to get the docstrings done tonight / tomorrow.

TomAugspurger commented:
line
====

- data
- x
- y
- use_index
- ax
- subplots
- sharex
- sharey
- layout
- grid
- title
- legend
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- colormap
- position
- table
- yerr
- xerr
- stacked
- sort_columns
- kwds


bar / barh
==========

- data
- x
- y
- use_index
- ax
- subplots
- sharex
- sharey
- layout
- grid
- title
- legend
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- colormap
- position
- table
- yerr
- xerr
- stacked
- sort_columns
- kwds

hist
====

- column
- by
- ~x~
- ~y~
- ~use_index~
- ax
- subplots
- sharex
- sharey
- layout
- grid
- title
- legend
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- colormap
- table
- stacked
- sort_columns
- kwds

box
===

- ax
- subplots
- sharex
- sharey
- layout
- grid
- title
- legend
- ~logy~ : isn't computed correctly.
- rot
- xlim
- ylim
- fontsize
- figsize
- table
- kwds

kde / density
=============

- ax
- subplots
- sharex
- sharey
- layout
- grid
- title
- legend
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- colormap
- colorbar
- table
- kwds

area
====

- x
- y
- use_index
- ax
- subplots
- sharex
- sharey
- layout
- grid
- title
- legend
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- table
- stacked
- sort_columns
- kwds

pie
===

- y
- ax
- subplots
- layout
- grid
- title
- legend
- rot
- xlim
- ylim
- fontsize
- figsize
- colormap
- table
- kwds

scatter
=======

- x
- y
- ax
- grid
- title
- ~legend~
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- colormap
- colorbar
- position
- table
- yerr
- xerr
- kwds

hexbin
======

- x
- y
- ax
- grid
- title
- ~legend~
- logx
- logy
- loglog
- xticks
- yticks
- rot
- xlim
- ylim
- fontsize
- secondary_y
- mark_right
- figsize
- style
- colormap
- colorbar
- position
- table
- yerr
- xerr
- gridsize
- C
- reduce_C_function
- kwds

TomAugspurger commented:
Unfortunately, the `__call__` signature is included in the docstring of the `.plot` method.

```python
In [9]: df.plot?
Type:        FramePlotMethods
String form: <pandas.tools.plotting.FramePlotMethods object at 0x1087b9860>
File:        /Users/tom/miniconda3/envs/py3/lib/python3.4/site-packages/pandas/pandas/tools/plotting.py
Definition:  df.plot(self, x=None, y=None, kind='line', ax=None, subplots=False, sharex=True, sharey=False, layout=None, figsize=None, use_index=True, title=None, grid=None, legend=True, style=None, logx=False, logy=False, loglog=False, xticks=None, yticks=None, xlim=None, ylim=None, rot=None, fontsize=None, colormap=None, table=False, yerr=None, xerr=None, secondary_y=False, sort_columns=False, **kwds)
Docstring:
DataFrameFrame plotting accessor

Examples
--------
>>> df.plot.line()
>>> df.plot.scatter('x', 'y')
>>> df.plot.hexbin()

The plotting methods can also be accessed by directly calling the
accessor with the ``kind`` argument:
``df.plot(kind='line')`` is equivalent to ``df.plot.line()``
Call def:    df.plot(x=None, y=None, kind='line', ax=None, subplots=False, sharex=True, sharey=False, layout=None, figsize=None, use_index=True, title=None, grid=None, legend=True, style=None, logx=False, logy=False, loglog=False, xticks=None, yticks=None, xlim=None, ylim=None, rot=None, fontsize=None, colormap=None, table=False, yerr=None, xerr=None, secondary_y=False, sort_columns=False, **kwds)
```

I'm not sure this is avoidable.

TomAugspurger commented:
I did some work on this a while back. I'll see how it looks and maybe open up a pull against your branch.

jreback commented:
@shoyer nice start....for 0.17.0!

shoyer commented:
@jreback @TomAugspurger Yes, let's get this in for 0.17. Even it's just our best guess for the full doc strings, we can fill in the long tail of options later.

shoyer commented:
OK, I have been updating this. To keep the complexity under control, so far all I've done is add the kind specific arguments to each plotting method. So at least it's clear which arguments are relevant for each kind, and these can be used positionally.

I would rather get this out in its somewhat incomplete form and add the long tail of plotting methods (as documented by @TomAugspurger above) later. That should be pretty straightforward at this point, if someone can figure out a way to consolidate the docstring and function signature generation without going insane.

Here's what the new API docs look like:

![screen shot 2015-05-29 at 10 43 22 am](https://cloud.githubusercontent.com/assets/1217238/7888755/a8bd20d4-05ef-11e5-958b-05729d949d91.png)

The main remaining blocker is that we get `alias of FramePlotMethods` as the API documentation for `DataFrame.plot`, without any docstring or generated page. @jorisvandenbossche do you know how to fix this? I can dig in if necessary.

jreback commented on doc/source/visualization.rst:
```diff
@@ -134,6 +135,9 @@ These include:
 * :ref:`'hexbin' <visualization.hexbin>` for hexagonal bin plots
 * :ref:`'pie' <visualization.pie>` for pie plots
 
+New in version 0.17, you can create these other plots using the method
```
add a ``..versionadded`` directive

shoyer replied:
I didn't realize that works on paragraphs

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -21,6 +21,15 @@ Check the :ref:`API Changes <whatsnew_0170.api>` and :ref:`deprecations <whatsne
+To alleviate this issue, we have added a new, optional plotting interface, which exposes each kind of plot as a method of the ``.plot`` attribute. Instead of writing ``series.plot(kind=<kind>, ...)``, you can now also use ``series.plot.<kind>(...)``.
+
+A major advantage of this approach is that required arguments for each plotting method now can appear first in the function signature, and can be referred to by position, e.g., you can write ``series.plot.pie('food')`` instead of ``series.plot(y='food', kind='pie')``.
+
```
couple of examples in a code-block maybe? (and maybe an ipython tab completion example)

shoyer replied:
Those are good ideas, I'll update, and probably add tab-completion to the plotting docs -- I had forgotten about that.

shoyer commented:
@jorisvandenbossche any ideas for fixing the doc generation? Even just a pointer to relevant docs would Sphinx would be helpful... my blind experiments did not have much luck.

jorisvandenbossche commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -21,6 +21,33 @@ Check the :ref:`API Changes <whatsnew_0170.api>` and :ref:`deprecations <whatsne
+To alleviate this issue, we have added a new, optional plotting interface, which exposes each kind of plot as a method of the ``.plot`` attribute. Instead of writing ``series.plot(kind=<kind>, ...)``, you can now also use ``series.plot.<kind>(...)``:
+
+.. ipython::
+	:verbatim:
```
You've got some tabs instead of spaces here

jorisvandenbossche commented on doc/source/api.rst:
```diff
@@ -651,12 +651,35 @@ the Categorical back to a numpy array, so levels and order information is not pr
+
+.. autosummary::
+   :toctree: generated/
+   :template: autosummary/accessor_plot.rst
```
Did you add this `accessor_plot.rst` file to the PR? I don't see it

jorisvandenbossche commented:
@shoyer I took a look at this, and I am afraid I don't directly see a way to solve this. 

The problem is that `DataFrame/Series.plot` is no longer a method, but a class.
So also when you do `df.plot?`, you get the class docstring. 
Just this fact is also already a bit annoying, as this will not show the signature as does `df.plot?` now.

To get the docstring page build for `Series.plot`, you have to use another template (see how it is done for cat/str/dt):

```
.. autosummary::
       :toctree: generated/
       :template: autosummary/accessor.rst

       Series.plot
```

But that does not solve the issue of above of the wrong docstring/wrong signature. The wrong docstring can of course easily be fixed by changing the docstring of the PlotMethods class. But the help of `df.plot?` will still look like that one of a class

shoyer commented:
Well, in the worst case scenario we could generate the rst on the API docs for .plot manually? That wouldn't be too bad

On Thu, Jun 25, 2015 at 5:29 AM, Joris Van den Bossche
<notifications@github.com> wrote:

> @shoyer I took a look at this, and I am afraid I don't directly see a way to solve this. 
> The problem is that `DataFrame/Series.plot` is no longer a method, but a class.
> So also when you do `df.plot?`, you get the class docstring. 
> Just this fact is also already a bit annoying, as this will not show the signature as does `df.plot?` now.
> To get the docstring page build for `Series.plot`, you have to use another template (see how it is done for cat/str/dt):
> ```
> .. autosummary::
>        :toctree: generated/
>        :template: autosummary/accessor.rst
>        Series.plot
> ```
> But that does not solve the issue of above of the wrong docstring/wrong signature. The wrong docstring can of course easily be fixed by changing the docstring of the PlotMethods class. But the help of `df.plot?` will still look like that one of a class
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/pydata/pandas/pull/9321#issuecomment-115234427

jorisvandenbossche commented:
The generation of the API pages won't be the problem I think. With my suggestion of above (using the correct template), I think this should work. 
Where I am more worried about is the `df.plot?` help in the console. We can make the docstring give the correct information, but it will still say it is a class and show no signature, we seems a potential source of confusion to me ..

jreback commented:
@shoyer status?

shoyer commented:
I think this just needs the fix to generate the right doc pages (and a rebate, of course). I'll need to experiment to figure that out, unless @jorisvandenbossche already know how...

I'm on vacation this week but will try to squeeze in some time :)

jreback commented:
status?

shoyer commented:
Ugh, it is a major struggle to get autodoc to handle `Series.plot` properly.

At build time (for the docs) `Series.plot` is a class object, so autodoc insists on using the class template for the entry on `api.html`. I can use a custom template to make the subpage look reasonable, but it's still not looking right in the API docs table of contents...

shoyer commented:
This still needs a major rebase, but I have finally succeeded in generating the right API docs!

--------------------------

Here's what the relevant sections look like on the main API page:

<img width="695" alt="screen shot 2015-09-05 at 12 14 23 pm" src="https://cloud.githubusercontent.com/assets/1217238/9701120/26eb256c-53d1-11e5-873c-6fd7b56635de.png">
<img width="674" alt="screen shot 2015-09-05 at 12 17 52 pm" src="https://cloud.githubusercontent.com/assets/1217238/9701124/37dc3ea6-53d1-11e5-8c59-7991542e9624.png">

-----------------------------

Here's the documentation page for `DataFrame.plot`:
<img width="682" alt="screen shot 2015-09-05 at 12 14 40 pm" src="https://cloud.githubusercontent.com/assets/1217238/9701122/2d1c6aae-53d1-11e5-9684-78486f9e1168.png">

-----------------------------

And here's an example of the doc pages for one of the submethods. Note that **`pandas.DataFrame.plot`** here is a link to the main `DataFrame.plot` page:
<img width="680" alt="screen shot 2015-09-05 at 12 15 01 pm" src="https://cloud.githubusercontent.com/assets/1217238/9701123/32e45fb4-53d1-11e5-8715-1e04f44e73c8.png">

jorisvandenbossche commented:
Nice hacks to get the autosummary working! :-) Glad you found a solution

jreback commented on pandas/tests/test_graphics.py:
```diff
@@ -473,8 +473,8 @@ def test_plot(self):
 
-        _check_plot_works(self.ts[:10].plot, kind='bar')
-        _check_plot_works(self.ts.plot, kind='area', stacked=False)
+        _check_plot_works(self.ts[:10].plot.bar)
```
maybe have a few duplicates where you use the plot accessor AND the function call (with ``kind``). You may have done this, just checking.

shoyer replied:
Done -- added explicit tests for all plot methods.

jreback commented:
lgtm. 

@sinhrks @TomAugspurger @jorisvandenbossche 

needs a rebase / squash

existing docs in visualization.rst to change to the new accessing methods? in a follow-up PR (pls make an issue)

JanSchulz commented on pandas/tools/plotting.py:
```diff
@@ -3388,6 +3401,392 @@ def _set_ticks_props(axes, xlabelsize=None, xrot=None,
+        -------
+        axes : matplotlib.AxesSubplot or np.array of them
+        """
+        return self(kind='line', **kwds)
```
You can get argument completition on newer python versions if you (programmatically on import) add a `__signature__` object: see this PR in mpl, which does it for decorators: https://github.com/matplotlib/matplotlib/pull/4829/files#diff-9ffb0d1db51a67ee4f37528e00715b3cR1597 

There is also a PR in IPython which would backport this to 2.7 and which mpl will use if IPython is installed.

shoyer replied:
Good to know. That's probably worth adding once we figure out all the other valid keyword argument...

jorisvandenbossche commented:
For the existing visualization.rst docs, I think it is OK to leave that for now

(also for later, question is one of the two we want to advertise the most)

shoyer committed:
ENH: plot method accessors

Fixes #9124

This PR adds plotting sub-methods like `df.plot.scatter()` as an alternative
to using `df.plot(kind='scatter')`.

I've added meaningful function signatures and documentation for a few of these
methods, but I would greatly appreciate help to fill in the rest -- this is a
lot of documentation to write! The entire point of this PR, of course, is to
have better introspection and docstrings.

shoyer commented:
Did the big rebase/squash -- let's see if things pass on Travis. Also added two followup issues for documentation. Assuming tests pass, this is ready for a final review.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -116,6 +116,35 @@ Releasing of the GIL could benefit an application that uses threads for user int
 .. _dask: https://dask.readthedocs.org/en/latest/
 .. _QT: https://wiki.python.org/moin/PyQt
 
+.. _whatsnew_0170.plot:
```
I would add to the highlites as pointer to this section.

shoyer replied:
done

jreback commented on pandas/core/frame.py:
```diff
@@ -5432,7 +5433,7 @@ def _put_str(s, space):
 
-DataFrame.plot = gfx.plot_frame
+DataFrame.plot = base.AccessorProperty(gfx.FramePlotMethods, gfx.FramePlotMethods)
 DataFrame.hist = gfx.hist_frame
```
hmm, maybe we should deprecate ``DataFrame.hist`` (and point to ``DataFrame.plot.hist``) as the canonical way?

TomAugspurger replied:
The defaults on `DataFrame.hist` are different that `DataFrame.plot.hist` which I think is why we have both for now. I could go either way on deprecating `DataFrame.hist`.

jreback commented on pandas/core/series.py:
```diff
@@ -2882,7 +2882,7 @@ def __init__(self, *args, **kwargs):
-Series.plot = _gfx.plot_series
+Series.plot = base.AccessorProperty(_gfx.SeriesPlotMethods, _gfx.SeriesPlotMethods)
 Series.hist = _gfx.hist_series
 
```
deprecate ``.hist`` ? (see above)

jreback commented:
lgtm. 

deprecate ``.hist``? (and point to ``.plot.hist``)
@jorisvandenbossche @sinhrks @TomAugspurger

TomAugspurger commented on doc/source/api.rst:
```diff
@@ -672,12 +672,34 @@ the Categorical back to a numpy array, so levels and order information is not pr
 ~~~~~~~~
 
+``Series.plot`` is both a callable method and a namespace attribute for
+specific plotting methods of form ``Series.plot.<kind>``.
```
"of form" -> "of the form". Same for DataFrame.plot below.

jorisvandenbossche commented:
I don't think a deprecation of `hist` is directly needed at the moment (they also don't have fully compatible signatures). But could certainly add a pointer in the hist docstring.

TomAugspurger commented:
:clap:

shoyer commented:
Added another followup issue for deprecating `DataFrame.hist`: https://github.com/pydata/pandas/issues/11053

I think this would be a good idea if we can sort out the API differences.

jreback commented:
ok, unless anyone has anything else, @shoyer merge on green.

jreback commented:
I don't the image in the whatsnew got commited: ``_static/whatsnew_plot_submethods.png``
can you push it up?

shoyer commented:
Oops. Just pushed to master: https://github.com/pydata/pandas/commit/eca9db97c2c7746ed8a599b81092be9735907a66

jreback commented:
awesome thanks!
